### PR TITLE
Fixes #3456 - Accept unicode in theme colors

### LINKED ISSUE #3456: Accented char in mybb_theme_colors.xml breaks default theme
Hello,

From years now, I translate mybb into french (installation included).

Since v1.8.0 a new functionnality was implemented : colors templates for default theme with the file /install/resources/mybb_themes_color.xml.
I translated the color themes names into french. So "Crepuscule" became "Crépucule". That file was created on february 2015 and never changed nor problematic.

But, since v1.8.16, this file causes an issue.
When editing theme settings, now I get the error "&#36;l['error_invalid_color']" defined in /admin/styles.lang.php, so I was unable to save anything.

Comparing "/admin/modules/style/themes.php" between 1.8.15 and 1.8.16, I see you restrict the stylesheets names to only alphanumeric characters in v1.8.16 and later!
Lines 1095 to 1107.
PHP Code:foreach(&#36;colors&nbsp;as&nbsp;&#36;color)    &nbsp;{   &nbsp;&#36;color&nbsp;=&nbsp;trim(&#36;color);   &nbsp;if(preg_match('/^[a-z0-9]+={1}[a-z0-9]+&#36;/i',&nbsp;&#36;color))    &nbsp;{          &nbsp;&#36;color&nbsp;=&nbsp;explode("=",&nbsp;&#36;color);          &nbsp;&#36;properties['colors'][&#36;color[0]]&nbsp;=&nbsp;&#36;color[1];   &nbsp;}   &nbsp;else   &nbsp;{         &nbsp;&#36;errors[]&nbsp;=&nbsp;&#36;lang-&gt;sprintf(&#36;lang-&gt;error_invalid_color,&nbsp;&#36;color);   &nbsp;}  &nbsp;}&nbsp;
So, now it's impossible made any changes in default theme. That'a pity!
If I change mybb_theme_colors.xml with new one without accented char, the error remains.
Rebuilding default_theme cache does not correct anything.

I tried to change 'é' to 'e' in database (mybb_themes and mybb_datacache), but doing this breaks completely the default theme ! The CSS are lost and now the error is :

Code:Warning [2] Invalid argument supplied for foreach() - Line: 1361
File: ...&#92;admin&#92;modules&#92;style&#92;themes.php PHP 5.6.12 (WINNT)
File     Line     Function
D:&#92;xampp&#92;htdocs&#92;mybb_test&#92;admin&#92;modules&#92;style&#92;themes.php     1361     errorHandler-&gt;error
D:&#92;xampp&#92;htdocs&#92;mybb_test&#92;admin&#92;index.php     821     require!
I can't understand why you changed this! Why a stylesheet name should not have accented chars?... I don't know who had the idea to change suddenly this behaviour, but my question is :
How can I correct this on a running mybb &gt;= 1.8.6 ?
I need to inform my members on how to correct without loosing all.

Note : I wrote a little script (see attachment) inspired from install index.php to re-execute the function insert_templates(). After uploaded the file in /install directory et replaced mybb_theme_colors.xml file, it seems working, but are there a best and/or light way to restore default theme functionnality after replaced mybb_theme_colors.xml file?

Thanks for answer.
[attachment=40848]

Original thread: [Accented char in mybb_theme_colors.xml breaks default theme](https://community.mybb.com/thread-219792.html)

### PR DESCRIPTION
**Pull request by Eldenroot**
Resolves #3456 

Thank you @lairdshaw for another constructive lesson!

### CODE REVIEW

File: admin/modules/style/themes.php [hunk 1]
```diff
@@ -1178,7 +1178,7 @@ function checkAction(id)
 			'properties' => $db->escape_string(my_serialize($properties))
 		);
 
-		// perform validation
+		// Perform validation
```
Comment (Sama34) [inline admin/modules/style/themes.php:1181]: Inline comments aren't documentation ;v

**Comment by Sama34** [comment]
Language string `error_invalid_color` should be updated to offer a more appropriate message.

**Review by Sama34 — CHANGES_REQUESTED** [review]

File: admin/modules/style/themes.php (hunk 1 above)
Comment (Eldenroot) [reply admin/modules/style/themes.php:1181]: They are not, but in this case when the file is modified it was stronger than me :)

**Comment by Eldenroot** [comment]
@Sama34 - I did small adjustment in the description. I am not a native speaker, so if we wanna improve the text, please let mek now what do you suggest. Thank you!

**Review by Sama34 — APPROVED** [review]

File: inc/languages/english/admin/style_themes.lang.php [hunk 2]
```diff
@@ -188,7 +188,7 @@
 $l['error_missing_stylesheet_extension'] = "This stylesheet must end with the correct file extension, for example, {1}<em>.css</em>";
 $l['error_invalid_parent_theme'] = "The selected parent theme does not exist. Please select a valid parent theme.";
 $l['error_invalid_templateset'] = "The selected template set does not exist. Please select a valid template set.";
-$l['error_invalid_color'] = "The defined color set \"{1}\" is either having invalid character(s) or not in prescribed format. Only alpha-numeric characters are allowed for naming separated by '='.";
+$l['error_invalid_color'] = "The defined color set \"{1}\" is either having invalid character(s) or not in prescribed format. Only numbers and unicode characters are allowed for naming separated by '='.";
```
Comment (lairdshaw) [inline inc/languages/english/admin/style_themes.lang.php]: Minor nitpick: maybe "Unicode letters" is more accurate than "unicode characters" - a letter is more specifically something you'd use in an ordinary word, whereas a character could be anything.

Comment (Sama34) [inline inc/languages/english/admin/style_themes.lang.php]:
```suggestion
$l['error_invalid_color'] = "The defined color set \"{1}\" is either having invalid character(s) or not in prescribed format. Only numbers and unicode letters are allowed for naming separated by '='.";
```

**Review by Sama34 — CHANGES_REQUESTED** [review]

File: inc/languages/english/admin/style_themes.lang.php (hunk 2 above)
Comment (lairdshaw) [reply inc/languages/english/admin/style_themes.lang.php]: One other nitpick: "is [...] having" is unnatural syntax to a native English speaker in this context. More natural would be "has". So, I'd suggest the first sentence be reworded to:

"The defined color set \\"{1}\\" either has invalid character(s) or is not in the prescribed format."

Perhaps, too, a comma after "naming" in the second sentence would be helpful for breaking the two phrases, plus a little more description, such as this:

"Only numbers, Unicode letters, and underscores are allowed in a color's name and its value, which must be separated by '='."

**Comment by Eldenroot** [comment]
Corrected. Thank you both for your valuable feedback!

**Review by Sama34 — APPROVED** [review]

**Comment by Eldenroot** [comment]
@Ben-MyBB - how to close this PR (it was merged yesterday)...

EDIT: Nevermind, Github has some issues apparently... it was displayed in opened PRs section... now it seems to be fine.

**Pull request merged**